loader/config.c

```diff
@@ -5,878 +5,876 @@
 {
-    oops("expected key %pSV to be unique!", &key);
-}
+    BUG_ON(offset >= cfg->size);
+    memzero(res, sizeof(struct find_result));
 
-NORETURN
-static void oops_on_unexpected_type(struct string_view key, enum value_type expected_mask, enum value_type got)
-{
-    struct string_view tstr_expected, tstr_got = value_type_as_str(got);
+    for (;;) {
+        struct config_entry *ent = &cfg->buffer[offset];
 
-    size_t types = __builtin_popcount(expected_mask);
-    size_t bit;
+        if (ent->t != CONFIG_ENTRY_VALUE)
+            break;
 
-    if (types == 1) {
-        tstr_expected = value_type_as_str(expected_mask);
+        if (sv_equals(ent->key, key)) {
+            res->last = offset;
 
-        oops("unexpected type for \"%pSV\"! expected: %pSV, got: %pSV\n",
-             &key, &tstr_expected, &tstr_got);
-    }
+            if (!res->count)
+                res->first = offset;
 
-    print_err("Oops!\nunexpected type for \"%pSV\"! expected one of: [", &key);
+            res->count++;
 
-    for (bit = 1; bit < VALUE_ANY; bit <<= 1) {
-        if (!is_type_set(expected_mask, bit))
-            continue;
+            if (res->count >= max)
+                break;
+        }
 
-        tstr_expected = value_type_as_str(bit);
-        print_err("%pSV", &tstr_expected);
-        if (--types == 0)
-            break;
+        if (ent->next == 0) break;
 
-        print_err(", ");
+        offset += ent->next;
     }
+}
 
-    print_err("] got: %pSV\n", &tstr_got);
-    for (;;);
+static bool is_of_type(enum value_type type, enum value_type mask)
+{
+    return (type & mask) == type;
 }
 
-static bool cfg_get_next_entry(struct config *cfg, size_t entry_offset, struct config_entry *this_entry,
-                               struct config_entry **out_entry, struct string_view *key)
+static bool cfg_find_ext(struct config *cfg, size_t offset, bool unique, struct string_view key, 
+                     enum value_type mask, struct value *val)
 {
-    for (;;) {
-        if (this_entry->offset_to_next_within_same_scope == 0)
-            return false;
+    BUG_ON(!mask);
+    BUG_ON(offset >= cfg->size);
 
-        entry_offset += this_entry->offset_to_next_within_same_scope;
-        this_entry = &cfg->buffer[entry_offset];
+    struct find_result res;
+    struct config_entry *ent;
 
-        if (key && !sv_equals(this_entry->key, *key))
-            continue;
+    cfg_find(cfg, offset == 0 ? 0 : offset + 1, key, 2, &res);
 
-        *out_entry = this_entry;
-        return true;
+    if (res.count > 1 && unique)
+        oops("%pSV must be unique", &key);
+
+    if (!res.count)
+        return false;
+
+    ent = &cfg->buffer[res.first];
+
+    if (!is_of_type(ent->as_value.type, mask)) {
+        struct string_view mistype = value_type_as_str(ent->as_value.type);
+        oops("%pSV has unexpected type of %pSV", &key, &mistype);
```

Comment:
This is actually a bit of a downgrade if you look at the old code. Previously it would log all the types set in `mask` so that it's easier for the user to understand what is expected. Can we get that functionality back? `oops_on_unexpected_type` could probably be copied verbatim (that is unless you want to improve it)